### PR TITLE
Support embedding Text test cases into Markdown files

### PR DESCRIPTION
**Pull request by vairamsvsjdo**
Support embedding Text test cases into Markdown files
fixes issue:  https://github.com/robotframework/robotframework/issues/2610

### CODE REVIEW

**Comment by vairamsvsjdo** [comment]
Navigating through legacy code constructs in robotframework: 
1. robotframework code several of them don't comply with PEP8 conventions: example, populators.py did not meet PEP8 and comment requirement.
2. where is utest for restreader? (I was looking for a sample to use for creating utest for mdreader.py)
3. Is there a contribution recommendations for using a file as test resource or use the complex route of html reader.? I am not sold on htmlreader test method. thoughts?

File: src/robot/parsing/mdreader.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+
+def MarkDownReader():
+    class MarkDownReader(object):
+        def __init__(self):
+            self.robot_lines = []
+            self.robot_data = ''
+
+        def robotize(self, md_file):
+            f = StringIO(md_file.read().decode('UTF-8'))
```
Comment (pekkaklarck) [inline src/robot/parsing/mdreader.py]: Wrapping the read data into a `StringIO` makes no sense. Much easier to just iterate over the lines of the original file.

File: src/robot/parsing/mdreader.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+                include_line = False
+                for line in f.readlines():
+                    if not include_line:
+                        include_line = (line.strip().lower() ==
+                                        "```robotframework")
+                    elif line.strip() == "```":
+                        include_line = False
+                    else:
+                        self.robot_lines.append(line)
+                self.robot_data = str(''.join(self.robot_lines))
```
Comment (pekkaklarck) [inline src/robot/parsing/mdreader.py]: Using `str` here is totally unnecessary. It also causes `UnicodeDecodeError` with Python 2 if the data happens to contain non-ASCII characters.

File: src/robot/parsing/populators.py [hunk 3]
```diff
@@ -28,11 +28,11 @@
 from .tsvreader import TsvReader
 from .txtreader import TxtReader
 from .restreader import RestReader
-
+from .mdreader import MarkDownReader
```
Comment (pekkaklarck) [inline src/robot/parsing/populators.py]: Should have two empty lines after imports.

File: src/robot/parsing/populators.py [hunk 4]
```diff
@@ -28,11 +28,11 @@
...
 
 READERS = {'html': HtmlReader, 'htm': HtmlReader, 'xhtml': HtmlReader,
-           'tsv': TsvReader , 'rst': RestReader, 'rest': RestReader,
```
Comment (pekkaklarck) [inline src/robot/parsing/populators.py]: Fixing small PEP-8 errors here in the code that you edit is fine, but in general it would be better to have them in separate PRs. Having lot of unrelated changes in a PR makes reviewing it harder.

File: src/robot/parsing/populators.py [hunk 5]
```diff
@@ -83,7 +83,7 @@ def start_table(self, header):
         self._populator.populate()
         table = self._datafile.start_table(DataRow(header).all)
         self._populator = self._populators[table.type](table) \
-                if table is not None else NullPopulator()
+            if table is not None else NullPopulator()
```
Comment (pekkaklarck) [inline src/robot/parsing/populators.py]: As I already commented, PEP-8 fixes in general belong to own PRs. In this particular case I also happen to like the original style better. Using four spaces for indentation when continuing a line is confusing.

**Comment by pekkaklarck** [comment]
Answers to your questions:

1. I'm sure there are PEP-8 violations still in the code and especially in unit tests. Fixing them is fine, but fixes to code that isn't otherwise edited by this PR should rather be moved into their own PR.

2. I'm not sure does RestReader have unit tests. It could be thoroughly tested with acceptance tests.

3. Could you clarify what you mean with the complex route in testing HtmlReader? Also, please read `CONTRIBUTING.rst` and `atest/README.rst` for information about testing in general.

File: src/robot/parsing/mdreader.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    # Introduction
+    This is an example markdown text
+    ## Script Example
+    ```robotframework
+    *** Test Cases ***
+    My Test
+        Log hello world from markdown    console=yes
+    ```
+    ## Some more documentation
+'''
```
Comment (vairamsvsjdo) [inline src/robot/parsing/mdreader.py]: agreed, will make appropriate change

File: src/robot/parsing/mdreader.py (hunk 1 above)
Comment (vairamsvsjdo) [inline src/robot/parsing/mdreader.py]: makes sense, will make appropriate change.

File: src/robot/parsing/populators.py (hunk 4 above)
Comment (vairamsvsjdo) [inline src/robot/parsing/populators.py]: agreed

File: src/robot/parsing/populators.py (hunk 5 above)
Comment (vairamsvsjdo) [inline src/robot/parsing/populators.py]: agreed

**Comment by vairamsvsjdo** [comment]
1. agreed
2. Its at least not there in the test parsing folder.
3. Please ignore that.

**Comment by vairamsvsjdo** [comment]
@pekkaklarck I was able to find atest for rest:
However, have these questions:

1. These tests seems to duplicate:
[rest](https://github.com/robotframework/robotframework/blob/master/atest/testdata/parsing/data_formats/rest/sample.rst)
[restdirective](https://github.com/robotframework/robotframework/blob/master/atest/testdata/parsing/data_formats/rest_directives/sample.rst)

2. As I was reviewing the atest test cases, I am realizing the volume of different use cases that markdown file format needs to support as equivalent to Rest. Does this pull request meet all those test cases or can this be as different features.

**Comment by vairamsvsjdo** [comment]
@pekkaklarck  how is [${RESTDIR}](https://github.com/robotframework/robotframework/blob/master/atest/robot/parsing/data_formats/rest.robot#L7) variable populated? Is it a global command line variable?
Where is the autobuild test script that runs when a PR is requested?

**Comment by vairamsvsjdo** [comment]
@pekkaklarck  made changes as requested

**Comment by pekkaklarck** [comment]
Great! Unfortunately I don't have time for a review in immediate future but I hopefully have time for this next week or week after that.

**Comment by vairamsvsjdo** [comment]
Great, thanks

**Comment by pekkaklarck** [comment]
Development hasn't been too active in the past, but we are starting RF 3.1 development and I'd like to get this enhancement in. On a quick look the approach used looks pretty good, but there are some places where code could be cleaned up a little. For example, imports weren't in format recommended by PEP-8. More importantly, some of the tests looked pretty strange with Robot data inside the code blocks formatted as tables like this:

```
=============  =========  ==============  ===============
  Settings       Value         Value           Value
=============  =========  ==============  ===============
 Default Tags   default1
￼=============  =========  ==============  ===============
```

My guess is that these are copied from the current reStructuredText tests, where this format is used when the idea is to convert reST files to HTML and parse them then. We are going to deprecate and remove the support for using HTML data (as well as reST what is converted to HTML), so supporting anything like that with Markdown doesn't make sense either. The current code doesn't seem to support this either, so I would assume it's enough to just remove these tests.

**Comment by vairamsvsjdo** [comment]
@pekkaklarck  This was assumed at the time of making it work in markdown, we decided in our implementation to use reST format such that it will enable us to see these in markdown with syntax highlighting and work as a document for Acceptance Testing.

Please provide how you envision and will make changes accordingly.
What formats are we looking to support?
How do you see a test script must look embedded with other documentation.

**Comment by pekkaklarck** [comment]
We got new parser in RF 3.2 (#3076) and this PR is incompatible with it. Support for Markdown would be a nice addition, though, and adding it on top of the new parser ought to be easier than it was earlier.
